### PR TITLE
feat(core): enable the use of timestamps

### PR DESCRIPTION
**Pull request by oed**
This PR does two things mainly:
### Puts timestamps in log entries for AnchorCommits
This is useful when you want to be able to know when certain updates happened and will be critical to determine if a signature done by a DID was done by a key that was valid at that point in time.

### Enables loading of documents using an `atTime` parameter
Both `loadDocument` and `multiQuery` now supports the `atTime` parameter. When specified Ceramic will load the AnchorCommit closest to that timestamp. If *atTime* is before any AnchorCommit the genesis commit will be loaded. Otherwise it will load the AnchorCommit which is has a timestamp smaller than, or equal to *atTime*

### CODE REVIEW

File: packages/common/src/doctype.ts [hunk 1]
```diff
@@ -123,14 +126,25 @@ export interface DocState {
  * behaviors that are performed as part of the operation.
  */
 export interface DocOpts {
-    // Whether or not to request an anchor after performing the operation.
+    /**
+     * Whether or not to request an anchor after performing the operation.
+     */
```
Comment (oed) [inline packages/common/src/doctype.ts:131]: This should put the comment into the generated docs!

File: packages/core/src/ceramic.ts [hunk 2]
```diff
@@ -630,11 +631,14 @@ class Ceramic implements CeramicApi {
...
 
     // If DocID is requested, return the document
-    if (docRef instanceof DocID) {
-      return doc
-    } else {
+    if (docRef instanceof CommitID) {
       // Here CommitID is requested, let's return document at specific commit
       return doc.rewind(docRef)
+    } else if (opts.atTime) {
+      const commitId = doc.findCommitAt(opts.atTime)
```
Comment (ukstv) [inline packages/core/src/ceramic.ts:638]: This, with `findCommitAt` do not belong here IMO, but let it live for now. Going to move it out of the document after the merge.

Comment (oed) [reply packages/core/src/ceramic.ts:638]: Ok, let me know where I should move it? Or do you want to later?

**Review by ukstv — APPROVED** [review]

File: packages/core/src/ceramic.ts (hunk 2 above)
Comment (ukstv) [reply packages/core/src/ceramic.ts:638]: We'll do it later. It's all about having snapshots, which seems to be pretty targeted change. For now let it live here.

File: packages/core/src/__tests__/ceramic-api.test.ts [hunk 3]
```diff
@@ -431,14 +433,26 @@ describe('Ceramic API', () => {
                    notDoc: '123' },
         metadata: { controllers: [controller] }
       })
+
+      // test data for the atTime feature
+      docFStates.push(docF.state)
+      docFTimestamps.push(Date.now())
```
Comment (stbrody) [inline packages/core/src/__tests__/ceramic-api.test.ts:439]: is this adding a timestamp to an update that was never anchored?

File: packages/core/src/document.ts [hunk 4]
```diff
@@ -271,6 +272,27 @@ export class Document implements DocStateHolder {
     this.subscriptionSet.add(subscription);
   }
 
+  /**
+   * Find the relevant AnchorCommit given a particular timestamp.
+   * Will return an AnchorCommit whos timestamp is earlier to or
```
Comment (stbrody) [inline packages/core/src/document.ts:277]:
```suggestion
   * Will return an AnchorCommit whose timestamp is earlier to or
```

**Review by stbrody — APPROVED** [review]

File: packages/core/src/__tests__/ceramic-api.test.ts (hunk 3 above)
Comment (oed) [reply packages/core/src/__tests__/ceramic-api.test.ts:439]: No this is just a timestamp for the points of times before and in between the actual anchor commits.

Comment (stbrody) [reply packages/core/src/__tests__/ceramic-api.test.ts:439]: Now I'm more confused.  I thought only AnchorCommits should have timestamps

Comment (oed) [reply packages/core/src/__tests__/ceramic-api.test.ts:439]: They do, but here I'm just capturing the time before and after each anchor commit timestamp.

**Pull request merged**